pkg/prometheus/promcfg.go

```diff
@@ -109,7 +130,131 @@ func generateConfig(p *v1alpha1.Prometheus, mons map[string]*v1alpha1.ServiceMon
 
 	return yaml.Marshal(cfg)
 }
+func generateServiceMonitorConfigSvc(m *v1alpha1.ServiceMonitor, svc v1.Service, i int, basicAuthSecrets map[string]BasicAuthCredentials) yaml.MapSlice {
+	cfg := yaml.MapSlice{
+		{
+			Key:   "job_name",
+			Value: fmt.Sprintf("%s/%s/%d", m.Namespace, m.Name, i),
+		},
+	}
+
+	//dnsConf := []yaml.MapSlice{
+	dnsConf := yaml.MapSlice{
+		{
+			Key: "names",
+			Value: []string{
+				svc.Spec.ExternalName,
+			},
+		},
+		{
+			Key:   "type",
+			Value: m.Spec.ExternalName.DnsType,
+		},
+		{
+			Key:   "refresh_interval",
+			Value: "30s",
+		},
+	}
+	//}
+
+	if m.Spec.ExternalName.DnsType != "SRV" {
+		portConf := yaml.MapItem{Key: "port", Value: 0}
+		for _, p := range svc.Spec.Ports {
+			if p.Name == m.Spec.ExternalName.Port {
+				portConf.Value = p.Port
+			}
+		}
+
+		if portConf.Value == 0 {
+			// At this point, we're not an SRV record, and we've not been explictly given a port.
+			// We require a port for this config to work.
+			// Fallback to the first port number on the service
+			portConf.Value = svc.Spec.Ports[0].Port
+		}
 
+		dnsConf = append(dnsConf, portConf)
+	}
+
+	cfg = append(cfg, yaml.MapItem{Key: "dns_sd_configs", Value: []yaml.MapSlice{dnsConf}})
+
+	if m.Spec.ExternalName.Interval != "" {
```

Comment:
From here on onwards is essentially duplicate logic of what is generated for the kubernetes endpoints discovery. You can just expect an Endpoints struct to be passed, then you just pass in `m.Spec.ExternalName.Endpoint` to get the right type, or just not embed the type but make it a field of the `ExternalName` struct.